eventstore/manager.go

```diff
@@ -0,0 +1,111 @@
+package eventstore
+
+import (
+	"io"
+
+	"github.com/google/uuid"
+	ds "github.com/ipfs/go-datastore"
+	kt "github.com/ipfs/go-datastore/keytransform"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/textileio/go-textile-core/threadservice"
+)
+
+var (
+	dsStoreManagerBaseKey = ds.NewKey("/manager")
+)
+
+type Manager struct {
+	io.Closer
+
+	config *StoreConfig
+
+	threadservice threadservice.Threadservice
+	stores        map[uuid.UUID]*Store
+}
+
+// NewManager hydrates stores from prefixes and start them.
+func NewManager(ts threadservice.Threadservice, opts ...StoreOption) (*Manager, error) {
+	config := &StoreConfig{}
+	for _, opt := range opts {
+		if err := opt(config); err != nil {
+			return nil, err
+		}
+	}
+
+	if config.Datastore == nil {
+		datastore, err := newDefaultDatastore(config.RepoPath)
+		if err != nil {
+			return nil, err
+		}
+		config.Datastore = datastore
+	}
+
+	// @todo: handle debug here or in store?
+
+	m := &Manager{
+		config:        config,
+		threadservice: ts,
+		stores:        make(map[uuid.UUID]*Store),
+	}
+
+	q, err := m.config.Datastore.Query(query.Query{
+		Prefix:   dsStoreManagerBaseKey.String(),
+		KeysOnly: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer q.Close()
+
+	for res := range q.Next() {
+		idStr := ds.RawKey(res.Key).Parent().Parent().Parent().Parent().Name() // reaching for the stars here
+		id, err := uuid.Parse(idStr)
+		if err != nil {
+			return nil, err
+		}
+		s, err := newStore(m.threadservice, m.config)
+		if err != nil {
+			return nil, err
+		}
+		s.datastore = kt.Wrap(s.datastore, kt.PrefixTransform{
```

Comment:
Maybe direct assignment of an unexported field may break some assumptions (not sure now, but may in the future).
If `New` fires a goroutine that does something before this line, would write to other datastore that would be leaked?

Maybe just do a copy of `m.config`, edit the `datastore` field, and then call `newStore(..)` with that cloned-modified config.

Reply:
nice, good call.. i'll make that change